sympy/matrices/matrices.py

```diff
@@ -3888,59 +3891,164 @@ def LUdecomposition(self, iszerofunc=_iszero):
-                # the element may not be strictly zero, the supplied iszerofunc's evaluation gave True
-                raise ValueError("No nonzero pivot found; inversion failed.")
-            if pivot != j:  # row must be swapped
-                A.row_swap(pivot, j)
-                p.append([pivot, j])
-            scale = 1 / A[j, j]
-            for i in range(j + 1, n):
-                A[i, j] = A[i, j] * scale
-        return A, p
+
+        if self.rows == 0 or self.cols == 0:
+            # Define LU decomposition of a matrix with no entries as a matrix
+            # of the same dimensions with all zero entries.
+            return self.zeros(self.rows, self.cols), []
+
+        lu = self.as_mutable()
+        row_swaps = []
+
+        pivot_col = 0
+        for pivot_row in range(0, lu.rows-1):
+
+            for candidate_pivot_col in range(pivot_col, self.cols):
+                for candidate_pivot_row in range(pivot_row, self.rows):
+                    iszeropivot = iszerofunc(lu[candidate_pivot_row,
+                                                candidate_pivot_col])
+                    if iszeropivot is None:
+                        iszeropivot = False
+                        break
+                    if not iszeropivot:
+                        break
+                if not iszeropivot:
+                    break
+            pivot_col = candidate_pivot_col
+
+            if rankcheck and pivot_col != pivot_row:
+                # All entries including and below the pivot position are
+                # zero, which indicates that the rank of the matrix is
+                # strictly less than min(num rows, num cols)
+                # Mimic behavior of previous implementation, by throwing a
+                # ValueError.
+                raise ValueError("Rank of matrix is strictly less than"
+                                 " number of rows or columns."
+                                 " Pass keyword argument"
+                                 " rankcheck=False to compute"
+                                 " the LU decomposition of this matrix.")
+
+            elif iszeropivot:
+                # iszeropivot is True after pivot search has completed
+                # indicates that there are no more entries to zero,
+                # so Gaussian elimination is complete.
+                return lu, row_swaps
+
+            if pivot_row != candidate_pivot_row:
+                # Row swap book keeping:
+                # Record which rows were swapped.
+                # Update stored portion of L factor by multiplying L on the
+                # left and right with the current permutation.
+                # Swap rows of U.
+                row_swaps.append([pivot_row, candidate_pivot_row])
+
+                # Update L.
+                for col in range(0, pivot_row):
+                    lu[pivot_row, col], lu[candidate_pivot_row, col] =\
+                        lu[candidate_pivot_row, col], lu[pivot_row, col]
```

Comment:
Not that I know of, but `U[r, i:] = c*U[r_pivot, i:] + U[r, i:]` seems clearer than the loop to me.